src/libfetchers/git.cc

```diff
@@ -5,32 +5,249 @@
+    return st.st_mtime + settings.tarballTtl > now;
+}
+
+bool touchCacheFile(const Path& path, const time_t& touch_time)
+{
+  struct timeval times[2];
+  times[0].tv_sec = touch_time;
+  times[0].tv_usec = 0;
+  times[1].tv_sec = touch_time;
+  times[1].tv_usec = 0;
+
+  return lutimes(path.c_str(), times) == 0;
+}
+
+Path getCachePath(std::string key)
+{
+    return getCacheDir() + "/nix/gitv3/" +
+        hashString(htSHA256, key).to_string(Base32, false);
+}
+
+// Returns the name of the HEAD branch.
+//
+// Returns the head branch name as reported by git ls-remote --symref, e.g., if
+// ls-remote returns the output below, "main" is returned based on the ref line.
+//
+//   ref: refs/heads/main       HEAD
+//   ...
+std::optional<std::string> readHead(const Path & path)
+{
+    auto [exit_code, output] = runProgram(RunOptions {
+        .program = "git",
+        .args = {"ls-remote", "--symref", path},
+    });
+    if (exit_code != 0) {
+      return std::nullopt;
+    }
+
+    // Matches the common case when HEAD points to a branch, e.g.:
+    //   "ref: refs/heads/main  HEAD".
+    const static std::regex head_ref_regex("^ref:\\s*([^\\s]+)\\s*HEAD$");
+    // Matches when HEAD points directly at a commit, e.g.:
+    //   "71abcd...  HEAD".
+    const static std::regex head_rev_regex("^([^\\s]+)\\s*HEAD$");
+
+    for (const auto & line : tokenizeString<std::vector<std::string>>(output, "\n")) {
+        std::smatch match;
+        if (std::regex_match(line, match, head_ref_regex)) {
+            debug("resolved HEAD ref '%s' for repo '%s'", match[1], path);
+            return match[1];
+        } else if (std::regex_match(line, match, head_rev_regex)) {
+            debug("resolved HEAD ref '%s' for repo '%s'", match[1], path);
+            return match[1];
+        }
+    }
+
+    return std::nullopt;
+}
+
+// Persist the HEAD ref from the remote repo in the local cached repo.
+bool storeCachedHead(const std::string& actualUrl, const std::string& headRef)
 {
-    return chomp(runProgram("git", true, { "-C", path, "rev-parse", "--abbrev-ref", "HEAD" }));
+    Path cacheDir = getCachePath(actualUrl);
+    try {
+        runProgram("git", true, { "-C", cacheDir, "symbolic-ref", "--", "HEAD", headRef });
+    } catch (ExecError &e) {
+        if (!WIFEXITED(e.status)) throw;
+        return false;
+    }
+    /* No need to touch refs/HEAD, because `git symbolic-ref` updates the mtime. */
+    return true;
 }
 
-static bool isNotDotGitDirectory(const Path & path)
+std::optional<std::string> readHeadCached(const std::string& actualUrl)
+{
+    // Create a cache path to store the branch of the HEAD ref. Append something
+    // in front of the URL to prevent collision with the repository itself.
+    Path cacheDir = getCachePath(actualUrl);
+    Path headRefFile = cacheDir + "/HEAD";
+
+    time_t now = time(0);
+    struct stat st;
+    std::optional<std::string> cachedRef;
+    if (stat(headRefFile.c_str(), &st) == 0) {
```

Comment:
```suggestion
    if (pathExists(headRefFile)) {
```

(Not 100% similar, but should do what we want)

Reply:
I think it's important to check the mtime in the conditional below. Otherwise we would cache HEAD indefinitely which seems like surprising behavior (e.g., HEAD branch is changed from master to main for some repo, and we will keep fetching whatever master points to).

Reply:
Oh yes, didn’t realize that `st` was used down the line.